Arduino/MiP_ProMini_Pack_Library/src/mip.cpp

```diff
@@ -2127,6 +2390,13 @@ void MiP::processOobResponseData(uint8_t commandByte)
     case MIP_CMD_GET_STATUS:
         length = 2;
         break;
+    case MIP_CMD_RECEIVE_IR_DONGLE_CODE:
+        // MIP_CMD_RECEIVE_IR_DONGLE_CODE is the only message delivered by MiP that has a
+        // variable length so we need to read the next byte which contains the length. 
+        highNibble = Serial.read();
+        lowNibble = Serial.read();
+        length = (parseHexDigit(highNibble) << 4) | parseHexDigit(lowNibble);
```

Comment:
ShouldFix: This code isn't very robust. What happens if the serial data is corrupt and the hi and low nibble aren't available or length comes back with a value greater than 4.

Reply:
Good point.  Here's what I'm thinking.

```
    case MIP_CMD_RECEIVE_IR_DONGLE_CODE:
        // MIP_CMD_RECEIVE_IR_DONGLE_CODE is the only message delivered by MiP that has a
        // variable length so we need to read the next byte which contains the length. 
        if(Serial.available() >= 2)
        {
            highNibble = Serial.read();
            lowNibble = Serial.read();
            length = (parseHexDigit(highNibble) << 4) | parseHexDigit(lowNibble);
            if(length < 0x02 || 0x04 < length)
            {
                uint8_t discardedBytes = discardUnexpectedSerialData();
                MiPStream.print(F("MiP: Bad IR code length: "));
                MiPStream.print(length, HEX);
                MiPStream.print(F(" (discarded "));
                MiPStream.print(discardedBytes);
                MiPStream.println(F(" bytes)"));               
            }
        }
        break;
```

Reply:
I don't have any Arduino/AVR hardware around here right now (I have parts to make more but too hot to do so right now) to test so I was wondering if you could verify that Serial.read() times out after some time or does it block waiting forever for data to arrive. The docs are a bit ambiguous on this.

The issue with calling Serial.available() here in the if statement is that if the data isn't immediately available then length will be left set to 0 and no error generated. It would be nice to at least wait around for a few milliseconds for the next couple of bytes to arrive.

Reply:
I think what you described with Serial.available() was right.  I got lots of bad OOB command bytes.   So then I tried while(Serial.available() < 2) {}  instead of the if statement and It looked like the numbers were right but shifted a single digit to the right. So that wasn't right either.  This seems to work:

```
    case MIP_CMD_RECEIVE_IR_DONGLE_CODE:
        // MIP_CMD_RECEIVE_IR_DONGLE_CODE is the only message delivered by MiP that has a
        // variable length so we need to read the next byte which contains the length. 
        if(Serial.available() < 2)
        {
            delay(50);
        }
        highNibble = Serial.read();
        lowNibble = Serial.read();
        length = (parseHexDigit(highNibble) << 4) | parseHexDigit(lowNibble);
        if(length < 0x02 || 0x04 < length)
        {
            uint8_t discardedBytes = discardUnexpectedSerialData();
            MiPStream.print(F("MiP: Bad IR code length: "));
            MiPStream.print(length, HEX);
            MiPStream.print(F(" (discarded "));
            MiPStream.print(discardedBytes);
            MiPStream.println(F(" bytes)"));               
        }
        break;
```

Going to test a few more sets of dongle code before I resubmit.